pkg/apis/pipeline/v1beta1/resultref_test.go

```diff
@@ -433,3 +434,343 @@ func TestLooksLikeResultRef(t *testing.T) {
+					Result:       "sumResult",
+				},
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			expressions, ok := tt.args.whenExpression.GetVarSubstitutionExpressions()
+			if !ok && tt.want != nil {
+				t.Fatalf("expected to find expressions but didn't find any")
+			} else {
+				got := v1beta1.NewResultRefs(expressions)
+				if d := cmp.Diff(tt.want, got); d != "" {
+					t.Errorf("TestNewResultReference/%s %s", tt.name, diff.PrintWantGot(d))
+				}
+			}
+		})
+	}
+}
+
+func TestHasResultReferenceWhenExpression(t *testing.T) {
+	type args struct {
+		whenExpression v1beta1.WhenExpression
+	}
+	tests := []struct {
+		name    string
+		args    args
+		wantRef []*v1beta1.ResultRef
+	}{
+		{
+			name: "Test valid expression",
+			args: args{
+				whenExpression: v1beta1.WhenExpression{
+					Input:    "sumResult",
+					Operator: selection.In,
+					Values:   []string{"$(tasks.sumTask.results.sumResult)"},
+				},
+			},
+			wantRef: []*v1beta1.ResultRef{
+				{
+					PipelineTask: "sumTask",
+					Result:       "sumResult",
+				},
+			},
+		}, {
+			name: "Test valid expression with dashes",
+			args: args{
+				whenExpression: v1beta1.WhenExpression{
+					Input:    "$(tasks.sum-task.results.sum-result)",
+					Operator: selection.In,
+					Values:   []string{"sum-result"},
+				},
+			},
+			wantRef: []*v1beta1.ResultRef{
+				{
+					PipelineTask: "sum-task",
+					Result:       "sum-result",
+				},
+			},
+		}, {
+			name: "Test valid expression with underscores",
+			args: args{
+				whenExpression: v1beta1.WhenExpression{
+					Input:    "$(tasks.sum-task.results.sum_result)",
+					Operator: selection.In,
+					Values:   []string{"sum-result"},
+				},
+			},
+			wantRef: []*v1beta1.ResultRef{
+				{
+					PipelineTask: "sum-task",
+					Result:       "sum_result",
+				},
+			},
+		}, {
+			name: "Test invalid expression: param substitution shouldn't be considered result ref",
+			args: args{
+				whenExpression: v1beta1.WhenExpression{
+					Input:    "$(params.paramName)",
+					Operator: selection.In,
+					Values:   []string{"sum-result"},
+				},
+			},
+			wantRef: nil,
+		}, {
+			name: "Test valid expression in array",
+			args: args{
+				whenExpression: v1beta1.WhenExpression{
+					Input:    "$sumResult",
+					Operator: selection.In,
+					Values:   []string{"$(tasks.sumTask.results.sumResult)", "$(tasks.sumTask2.results.sumResult2)"},
+				},
+			},
+			wantRef: []*v1beta1.ResultRef{
+				{
+					PipelineTask: "sumTask",
+					Result:       "sumResult",
+				},
+				{
+					PipelineTask: "sumTask2",
+					Result:       "sumResult2",
+				},
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			expressions, ok := tt.args.whenExpression.GetVarSubstitutionExpressions()
+			if !ok {
+				t.Fatalf("expected to find expressions but didn't find any")
+			}
+			got := v1beta1.NewResultRefs(expressions)
+			sort.Slice(got, func(i, j int) bool {
```

Comment:
might be worth a comment to explain why the order is non deterministic (do we want it to be non deterministic?)